Apply performList && performGet if applicable.

bryanhelmig opened the pull request:
I like this but I'm not 100% sold on this - let's chat about this @bcooksey. The alternative is that the backend introspects across resources and triggers instead of having it right there.

bryanhelmig committed:
apply list.operation.perform to hook.operation.performList if present

bryanhelmig committed:
fix resource.list.operation typo

bryanhelmig committed:
fix other typo as well

bryanhelmig committed:
slip in pending schema change for performGet as well

BrunoBernardino reviewed (approved):
I think this looks good though it might not seem intuitive for us at a glance, because of the new "resources" concept.

bcooksey committed:
Ensure we snag performList and performGet from resource for independent triggers/searches/creates

Also fixed a bug where we skipped output fields and samples

bcooksey committed:
Turn a schema tool function private

Callers really need to go through compile app, not that converter

bcooksey committed:
Make sure creating the trigger from a resource does not modify the resource

bcooksey committed:
Nuke two unneeded deep copys

The first is not needed because compileApp is passing a copy in already. The second is not needed because we deep copy each individual resource to build the trigger/search/create, so caller can modify at will

bcooksey committed:
Refactor how apps are compiled so that some code can be reused

bcooksey committed:
Do not overwrite explicitly defined items on an app with generated one from a resource

This is more consistent with everything else we do around resources, where you can define a property like performGet or outputFields in your operation and we will not overwrite it from the linked resource

bcooksey commented:
Ok, this should be good for final review. As I was working on the `performList` and `performGet` copying from resources, I realized that we were doing it inconsistently for other fields that the schema claims we should be, so those are fixed.

And yes, I slipped some data copying tweaks in that were really out of scope for this PR, but they were small and I didn't want to forget.

bcooksey committed:
Update changelog

BrunoBernardino commented on src/tools/schema.js:
```diff
@@ -46,23 +41,59 @@ const convertResourceDos = (appRaw) => {
+    const resource = appRaw.resources[action.operation.resource];
+
+    if (type === 'trigger' && action.operation.type === 'hook') {
+      if (resource.list && resource.list.operation && resource.list.operation.perform) {
```
Instead of this long line we can use `lodash` and `if (_.get(resource, 'list.operation.perform')) {` which will return undefined if any of the properties in that "path" don't exist.

Same for similar code below.

BrunoBernardino commented on src/tools/schema.js:
```diff
@@ -46,23 +41,59 @@ const convertResourceDos = (appRaw) => {
+ */
+const copyPropertiesFromResource = (type, action, appRaw) => {
+  if (appRaw.resources && action.operation && appRaw.resources[action.operation.resource]) {
+    const copyableProperties = ['outputFields', 'sample'];
```
I thought the correct name to be `copiableProperties`? I _think_ it's a typo, which I'd love to see fixed, even if it's not important 😬

bcooksey replied:
I think this is a US vs. UK spelling. I'm finding `copyable` in some online dictionaries, but not the Oxford English Dictionary.

BrunoBernardino replied:
Wow, interesting! Looks so strange to me 😄

BrunoBernardino reviewed (approved):
Love it. It looks great! Just a comment to improve readability of some code, a typo, and that's it!

bcooksey committed:
Refactor a bit of code to be less verbose
